doc/util/collect-syntax.py

```diff
@@ -0,0 +1,327 @@
+#!/usr/bin/env python3
+# encoding: utf-8
+
+
+"""
+collect syntax
+"""
+
+
+from __future__ import print_function
+import argparse
+import os
+import re
+import sys
+import stat
+import traceback
+
+
+# Regular expressions to identify productions within RST files.
+_match_syntax_begin = re.compile(r"\s*.. code-block:: syntax\s*\n")
+_match_syntax_end = re.compile(r"\S+")
+_match_production = re.compile(r" {3}(?P<ident>([a-zA-Z0-9._-]+)):.*")
+_match_expansion = re.compile(r" {3} +")
```

Comment:
Is this really looking for 3 spaces? That seems a bit too fragile to me. In `doc/util/extract-rst-tests.py` I added a function that knows how to read an rst block (based on indenting). You could use that here I think. It seems to me we could/should use `read_block` or something like it both for the `code-block:: syntax` and also for the block from e.g. `expression:`.

Reply:
I will look into it, but as far as I know every single syntax block is indented 3 spaces on the first level.

Reply:
Right, but that doesn't mean the tooling should break if that stops being true in the future.

Reply:
Only thing left for me to fix now is this.+ the global variable names (if we decide we have a problem with them).